## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('../Challenge 4/Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe

<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#df
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [5]:
# What three item categories had the most entries?
df_count = df['category'].value_counts().head(3)
df_count

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?
df_count_sub = df['subcategory'].value_counts().head(1)
df_count_sub

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
df_count_sub = df['client_id'].value_counts().head(5)
df_count_sub

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
df_count_sub_top5 = df_count_sub.index.tolist()
df_count_sub_top5


[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the client with the most entries order order?
df_qty = df.groupby(["client_id"])[["qty"]].agg('sum')
df_qty.sort_values(by=["qty","client_id", ], ascending=False).head(5)

qty
client_id         
78965      4033302
86721      1010426
30143       846069
85931       644423
77354       506382

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['Subtotal'] = df['unit_price'] * df['qty']
df 


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   Subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
#unit_price	unit_weight	qty subtotal
#price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

def shipping_price(scost):
    weight = scost['unit_weight']
    quantity = scost['qty']
    total_weight = weight * quantity

    if total_weight > 50:
        return total_weight * 7
    else:
        return total_weight * 10
    
df['Shipping Price'] = df.apply(shipping_price, axis=1)

# "Apply" the function to the amount column
#donors_df['Match Amount'] = donors_df['Amount'].apply(match_amount)
#donors_df.head()
df['Amount'] = df['Shipping Price'] + df['Subtotal']
df     


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number   Subtotal  \
0        1096.80     762.71         7.50   105            1  115164.00   
1          24.95      15.09         1.49    21            0     523.95   
2          13.52       7.86         1.68    39            6     527.28   
3          36.42      24.85         1.23    29            3    1056.18   
4         195.10     108.17        46.43    20            1    3902.00   
...          ...        ...          ...   ...          ...        ...   
54634      83.13      51.60         2.25    33            8    2743.29   
54635     206.59     175.46        11.70    47            0    9709.73   
54636      65.66      57.31         4.16   475            6   31188.50   
54637       1.48       1.22        18.04   112            0     165.76   
54638       3.01       1.61         2.07  1031            5    3103.31   

       Shipping Price     Amount  
0             5512.50  120676.50  
1              312.90     836.85  
2              458.64     985.92  
3              356.70    1412.88  
4             6500.20   10402.20  
...               ...        ...  
54634          519.75    3263.04  
54635         3849.30   13559.03  
54636        13832.00   45020.50  
54637        14143.36   14309.12  
54638        14939.19 

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
tax_t = df['Amount']*.0925 # Beware of the 0 
Total_Price = tax_t + df['Amount']
df['Total Price'] = Total_Price

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number   Subtotal  Shipping Price     Amount  \
0             7.50   105            1  115164.00         5512.50  120676.50   
1             1.49    21            0     523.95          312.90     836.85   
2             1.68    39            6     527.28          458.64     985.92   
3             1.23    29            3    1056.18          356.70    1412.88   
4            46.43    20            1    3902.00         6500.20   10402.20   
...            ...   ...          ...        ...             ...        ...   
54634         2.25    33            8    2743.29          519.75    3263.04   
54635        11.70    47            0    9709.73         3849.30   13559.03   
54636         4.16   475            6   31188.50        13832.00   45020.50   
54637        18.04   112            0     165.76        14143.36   14309.12   
54638         2.07  1031            5    3103.31        14939.19   18042.50   

         Total Price  
0      131839.076250  
1         914.258625  
2        1077.117600  
3        1543.571400  
4       11364.403500  
...              ...  
54634    3564.871200  
54635   14813.240275  
54636   49184.896250  
54637   15632.713600 

In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
 
seller_cost = df['unit_cost'] * df['qty']

df['Sell Cost'] = seller_cost + df['Shipping Price']
df



first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   Subtotal  Shipping Price     Amount    Total Price  \
0                1  115164.00         5512.50  120676.50  131839.076250   
1                0     523.95          312.90     836.85     914.258625   
2                6     527.28          458.64     985.92    1077.117600   
3                3    1056.18          356.70    1412.88    1543.571400   
4                1    3902.00         6500.20   10402.20   11364.403500   
...            ...        ...             ...        ...            ...   
54634            8    2743.29          519.75    3263.04    3564.871200   
54635            0    9709.73         3849.30   13559.03   14813.240275   
54636            6   31188.50        13832.00   45020.50   49184.896250   
54637            0     165.76        14143.36   14309.12   15632.713600   
54638            5    3103.31        14939.19   18042.50   19711.431250   

       Sell Cost  
0       85597.05  
1         629.79  
2         765.18  
3        1077.35  
4        8663.60  
...          ...  
54634    2222.55  
54635   12095.92  
54636   41054.25  
54637   14280.00  
54638   16599.10  

[54639 rows x 23 columns]

In [14]:
# Create a column for the profit of each line using line cost and line price

df['Profit'] = df['Sell Cost'] - df['Subtotal']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

        Subtotal  Shipping Price     Amount    Total Price  Sell Cost  \
0      115164.00         5512.50  120676.50  131839.076250   85597.05   
1         523.95          312.90     836.85     914.258625     629.79   
2         527.28          458.64     985.92    1077.117600     765.18   
3        1056.18          356.70    1412.88    1543.571400    1077.35   
4        3902.00         6500.20   10402.20   11364.403500    8663.60   
...          ...             ...        ...            ...        ...   
54634    2743.29          519.75    3263.04    3564.871200    2222.55   
54635    9709.73         3849.30   13559.03   14813.240275   12095.92   
54636   31188.50        13832.00   45020.50   49184.896250   41054.25   
54637     165.76        14143.36   14309.12   15632.713600   14280.00   
54638    3103.31        14939.19   18042.50   19711.431250   16599.10   

         Profit  
0     -29566.95  
1        105.84  
2        237.90  
3         21.17  
4       4761.60  
...         ...  
54634   -520.74  
54635   2386.19  
54636   9865.75  
54637  14114.24  
54638  13495.79  

[54639 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
dicts = [{'Order': '2742071', 'total:': '152811.89'},
        {'Order': '2173913', 'total:': '162388.71'},
        {"Order": '6128929','total:': '923441.25'}]

dicts_df = pd.DataFrame(dicts)
dicts_df

Order     total:
0  2742071  152811.89
1  2173913  162388.71
2  6128929  923441.25

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
#1st part of my flaot format for Total Price
df['Total Price'] = df['Total Price'].astype(float)

sum_df = df[["client_id", "Total Price"]]
#2nd part of my flaot format for Total Price
pd.options.display.float_format = '{:.2f}'.format

sum_df.sort_values(by=["client_id", "Total Price"], ascending=False).head(5)


client_id  Total Price
35379      99984  16257453.17
42337      99984   1074078.87
4895       99984    501301.35
50091      99984    244601.04
43411      99984    161957.92

In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
df['Total Price'] = df['Total Price'].astype(float)

sum_df = df[["client_id", "qty", "Shipping Price", "Total Price", "Profit" ]]

pd.options.display.float_format = '{:.2f}'.format

sum_df.sort_values(by=["client_id", "qty", "Shipping Price", "Total Price", "Profit"], ascending=False).head(5)

client_id    qty  Shipping Price  Total Price      Profit
35379      99984  33950      2395512.00  16257453.17 -2092338.50
42337      99984   6145       360035.55   1074078.87   194857.95
43411      99984   2067        53824.68    161957.92    17342.13
958        99984   1647        25594.38     43562.20    21740.40
4895       99984    973        12668.46    501301.35  -136589.74

In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 

# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
df['Total Price'] = df['Total Price'].astype(float)
df['Profit'] = df['Profit'].astype(float)
df['Shipping Price'] = df['Shipping Price'].astype(float)

df['Updated Total Price'] = df['Total Price'].apply(lambda x: f"${x:,.2f}")
df['Updated Profit'] = df['Profit'].apply(lambda x: f"${x:,.2f}")
df['Updated Shipping Price'] = df['Shipping Price'].apply(lambda x: f"${x:,.2f}")

sum_df = df[["client_id", "qty", "Updated Shipping Price", "Updated Total Price", "Updated Profit" ]]

sum_df.sort_values(by=["client_id", "qty", "Updated Shipping Price", "Updated Total Price", "Updated Profit"], ascending=False).head(5)




client_id    qty Updated Shipping Price Updated Total Price  \
35379      99984  33950          $2,395,512.00      $16,257,453.17   
42337      99984   6145            $360,035.55       $1,074,078.87   
43411      99984   2067             $53,824.68         $161,957.92   
958        99984   1647             $25,594.38          $43,562.20   
4895       99984    973             $12,668.46         $501,301.35   

       Updated Profit  
35379  $-2,092,338.50  
42337     $194,857.95  
43411      $17,342.13  
958        $21,740.40  
4895     $-136,589.74

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
df['Total Price'] = df['Total Price'].astype(float)
df['Profit'] = df['Profit'].astype(float)
df['Shipping Price'] = df['Shipping Price'].astype(float)

df['Updated Total Price'] = df['Total Price'].apply(lambda x: f"${x:,.2f}")
df['Updated Profit'] = df['Profit'].apply(lambda x: f"${x:,.2f}")
df['Updated Shipping Price'] = df['Shipping Price'].apply(lambda x: f"${x:,.2f}")

sum_df = df[["client_id", "qty", "Updated Shipping Price", "Updated Total Price", "Updated Profit" ]]

sum_df.sort_values(by=["client_id", "qty", "Updated Shipping Price", "Updated Total Price", "Updated Profit"], ascending=False).head(5)


client_id    qty Updated Shipping Price Updated Total Price  \
35379      99984  33950          $2,395,512.00      $16,257,453.17   
42337      99984   6145            $360,035.55       $1,074,078.87   
43411      99984   2067             $53,824.68         $161,957.92   
958        99984   1647             $25,594.38          $43,562.20   
4895       99984    973             $12,668.46         $501,301.35   

       Updated Profit  
35379  $-2,092,338.50  
42337     $194,857.95  
43411      $17,342.13  
958        $21,740.40  
4895     $-136,589.74